File: src/Bionic.Bridge.Capacitor/FilesystemBridge.cs
using System.Threading.Tasks;
using Microsoft.JSInterop;

namespace Bionic.Bridge.Capacitor {
    public static class FilesystemBridge {
        private const string FilesystemAppendFileFunctionName = "BionicBridge.Capacitor.Filesystem.appendFile";
        private const string FilesystemDeleteFileFunctionName = "BionicBridge.Capacitor.Filesystem.deleteFile";
        private const string FilesystemGetUriFunctionName = "BionicBridge.Capacitor.Filesystem.getUri";
        private const string FilesystemMkdirFileFunctionName = "BionicBridge.Capacitor.Filesystem.mkdir";
        private const string FilesystemReadDirFunctionName = "BionicBridge.Capacitor.Filesystem.readdir";
        private const string FilesystemReadFileFunctionName = "BionicBridge.Capacitor.Filesystem.readFile";
        private const string FilesystemRmdirFileFunctionName = "BionicBridge.Capacitor.Filesystem.rmdir";
        private const string FilesystemStatFunctionName = "BionicBridge.Capacitor.Filesystem.stat";
        private const string FilesystemWriteFileFunctionName = "BionicBridge.Capacitor.Filesystem.writeFile";

        public static Task<FileAppendResult> AppendFile(FileAppendOptions options) =>
            JSRuntime.Current.InvokeAsync<FileAppendResult>(FilesystemAppendFileFunctionName, options);

        public static Task<FileDeleteResult> DeleteFile(FileDeleteOptions options) =>
            JSRuntime.Current.InvokeAsync<FileDeleteResult>(FilesystemDeleteFileFunctionName, options);

        public static Task<GetUriResult> GetUri(GetUriOptions options) =>
            JSRuntime.Current.InvokeAsync<GetUriResult>(FilesystemGetUriFunctionName, options);

        public static Task<MkdirResult> Mkdir(MkdirOptions options) =>
            JSRuntime.Current.InvokeAsync<MkdirResult>(FilesystemMkdirFileFunctionName, options);

        public static Task<FileReadResult> ReadFile(FileReadOptions options) =>
            JSRuntime.Current.InvokeAsync<FileReadResult>(FilesystemReadFileFunctionName, options);

        public static Task<ReaddirResult> ReadDir(ReaddirOptions options) =>
            JSRuntime.Current.InvokeAsync<ReaddirResult>(FilesystemReadDirFunctionName, options);

        public static Task<RmdirResult> Rmdir(RmdirOptions options) =>
            JSRuntime.Current.InvokeAsync<RmdirResult>(FilesystemRmdirFileFunctionName, options);

        public static Task<StatResult> Stat(StatOptions options) =>
            JSRuntime.Current.InvokeAsync<StatResult>(FilesystemStatFunctionName, options);

        public static Task<FileWriteResult> WriteFile(FileWriteOptions options) =>
            JSRuntime.Current.InvokeAsync<FileWriteResult>(FilesystemWriteFileFunctionName, options);
    }

    public class FileAppendOptions: FileWriteOptions {}

    public class FileWriteOptions: FileReadOptions {
        public string data;
    }

    public class FileReadOptions: FileDeleteOptions {
        public FilesystemEncoding encoding;
    }

    public class GetUriOptions: FileDeleteOptions { }
    public class ReaddirOptions: FileDeleteOptions { }
    public class RmdirOptions: FileDeleteOptions { }
    public class StatOptions: FileDeleteOptions { }

    public class FileDeleteOptions {
        public string path;
        public FilesystemDirectory directory;
    }

    public class MkdirOptions : FileDeleteOptions {
        public bool createIntermediateDirectories;
    }

    public class GetUriResult {
        public string uri;
    }

    public class FileAppendResult { }

    public class FileDeleteResult { }

    public class MkdirResult { }

    public class FileWriteResult { }

    public class FileReadResult {
        public string data;
    }

    public class ReaddirResult {
        public string[] files;
    }

    public class RmdirResult { }

    public class StatResult {
        public string type;
        public long size;
        public long ctime;
        public long mtime;
        public string uri;
    }

    public enum FilesystemDirectory {
        Application,
        Documents,
        Data,
        Cache,
        External,
        ExternalStorage
    }

    public enum FilesystemEncoding {
        UTF8,
        ASCII,
        UTF16,
    }
}

File: src/Bionic.Bridge.Capacitor.JS/src/FilesystemBridge.ts
import {
  FileAppendOptions,
  FileAppendResult,
  FileDeleteOptions,
  FileDeleteResult,
  FileReadOptions,
  FileReadResult,
  FilesystemDirectory,
  FilesystemEncoding,
  FileWriteOptions,
  FileWriteResult,
  GetUriOptions,
  GetUriResult,
  MkdirOptions,
  MkdirResult,
  Plugins,
  ReaddirOptions,
  ReaddirResult,
  RmdirOptions,
  RmdirResult,
  StatOptions,
  StatResult
} from "@capacitor/core";

const {Filesystem} = Plugins;

interface IFilesystem {
  appendFile(options: FileAppendOptions): Promise<FileAppendResult>;

  deleteFile(options: FileDeleteOptions): Promise<FileDeleteResult>;

  getUri(options: GetUriOptions): Promise<GetUriResult>;

  mkdir(options: MkdirOptions): Promise<MkdirResult>;

  readFile(options: FileReadOptions): Promise<FileReadResult>;

  readdir(options: ReaddirOptions): Promise<ReaddirResult>;

  rmdir(options: RmdirOptions): Promise<RmdirResult>;

  stat(options: StatOptions): Promise<StatResult>;

  writeFile(options: FileWriteOptions): Promise<FileWriteResult>;
}

export class FilesystemBridge implements IFilesystem {
  async appendFile(options: FileAppendOptions): Promise<FileAppendResult> {
    if (!Filesystem) return Promise.reject("Filesystem not available.");
    options.encoding = FilesystemBridge.translateEncode(options.encoding);
    options.directory = FilesystemBridge.translateDirectory(options.directory);
    return await Filesystem.appendFile(options);
  }

  async deleteFile(options: FileDeleteOptions): Promise<FileDeleteResult> {
    if (!Filesystem) return Promise.reject("Filesystem not available.");
    options.directory = FilesystemBridge.translateDirectory(options.directory);
    return await Filesystem.deleteFile(options);
  }

  async getUri(options: GetUriOptions): Promise<GetUriResult> {
    if (!Filesystem) return Promise.reject("Filesystem not available.");
    options.directory = FilesystemBridge.translateDirectory(options.directory);
    return await Filesystem.getUri(options);
  }

  async mkdir(options: MkdirOptions): Promise<MkdirResult> {
    if (!Filesystem) return Promise.reject("Filesystem not available.");
    options.directory = FilesystemBridge.translateDirectory(options.directory);
    return await Filesystem.mkdir(options);
  }

  async readFile(options: FileReadOptions): Promise<FileReadResult> {
    if (!Filesystem) return Promise.reject("Filesystem not available.");
    options.encoding = FilesystemBridge.translateEncode(options.encoding);
    options.directory = FilesystemBridge.translateDirectory(options.directory);
    return await Filesystem.readFile(options);
  }

  async readdir(options: ReaddirOptions): Promise<ReaddirResult> {
    if (!Filesystem) return Promise.reject("Filesystem not available.");
    options.directory = FilesystemBridge.translateDirectory(options.directory);
    return await Filesystem.readdir(options);
  }

  async rmdir(options: RmdirOptions): Promise<RmdirResult> {
    if (!Filesystem) return Promise.reject("Filesystem not available.");
    options.directory = FilesystemBridge.translateDirectory(options.directory);
    return await Filesystem.rmdir(options);
  }

  async stat(options: StatOptions): Promise<StatResult> {
    if (!Filesystem) return Promise.reject("Filesystem not available.");
    options.directory = FilesystemBridge.translateDirectory(options.directory);
    return await Filesystem.stat(options);
  }

  async writeFile(options: FileWriteOptions): Promise<FileWriteResult> {
    console.log(options);
    if (!Filesystem) return Promise.reject("Filesystem not available.");
    options.encoding = FilesystemBridge.translateEncode(options.encoding);
    options.directory = FilesystemBridge.translateDirectory(options.directory);
    return await Filesystem.writeFile(options);
  }

  private static translateEncode(encode: any): FilesystemEncoding {
    switch (encode) {
      case 1:
        return FilesystemEncoding.ASCII;
      case 2:
        return FilesystemEncoding.UTF16;
      default:
        return FilesystemEncoding.UTF8;
    }
  }

  private static translateDirectory(directory: any): FilesystemDirectory {
    switch (directory) {
      case 1:
        return FilesystemDirectory.Documents;
      case 2:
        return FilesystemDirectory.Data;
      case 3:
        return FilesystemDirectory.Cache;
      case 4:
        return FilesystemDirectory.External;
      case 5:
        return FilesystemDirectory.ExternalStorage;
      default:
        return FilesystemDirectory.Documents;
    }
  }
}
